### ISSUE TITLE
WindowServer: ASSERTION FAILED: screen terminates `WindowServer` depending on where you move your cursor

### ISSUE THREAD

**Issue by kemzeb** [SerenityOS/serenity#18668]
It appears that `WindowServer` has recently begun to crash due to an assertion failure in /Userland/Services/WindowServer/Screen.cpp:46. Depending on where you move your cursor, this assertion will fail and will cause the crash. The surest place where you can reproduce (at least for me) is when you move your cursor to the bottom-left corner of the screen.

**Comment by supercomputer7** [comment]
I was able to reproduce this when "violently" moving the cursor attached to a moving window, around from a corner to another corner.

**Comment by BenWiederhake** [comment]
Heh, funny, I just discovered the same bug and wanted to create a new issue. Here's a screen recording:

https://user-images.githubusercontent.com/2690845/236641358-f109ae98-618f-4fa5-8efe-ebd3cef63d6d.mp4

I added a `dbgln()` just before the assert, to demonstrate that the coordinates really are out of bounds.

**Comment by BenWiederhake** [comment]
Root cause seems to be a difference in opinion what exactly `constrain` is supposed to mean:

In `Screen.cpp`:
```
    if (!moved_to_screen) {
        m_cursor_location = m_cursor_location.constrained(current_screen.rect());
```
The assumption here seems to be that `m_cursor_location` is now a valid location inside the `current_screen`.

However, in `Point.cpp`:
```
void Point<T>::constrain(Rect<T> const& rect)
{
    m_x = AK::clamp<T>(x(), rect.left(), rect.left() + rect.width());
```
This does not hold due to fencepost error, as there are `rect.width() + 1` pixels in the range `rect.left()` through `rect.left() + rect.width()`. I'll make a PR that adjusts `Point::constrain`, since it really should mean "inside".

**Comment by BenWiederhake** [comment]
Fixed by #18687
